### ISSUE TITLE
camel case warning on impl trait

### ISSUE THREAD

**Issue by hellow554** [rust-lang-nursery/rust-clippy#2314]
I guess this should not happen :)

```
warning: type parameter `impl Iterator<Item = &'a Port>` should have a camel case name such as `Impl iterator<item = &'a port>`
   --> src/main.rs:108:44
    |
108 | fn do_things<'a>(state: &Port, candidates: impl Iterator<Item=&'a Port>) {
    |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: #[warn(non_camel_case_types)] on by default
```

**Comment by phansch** [comment]
I'm new to rust and especially clippy and was looking through the beginner issues here, so apologies if I got this wrong. As far as I can tell, this lint is defined in [`rustc`](https://github.com/rust-lang/rust/blob/8a11b8cdd7cc7214c2bb9e77a6c961dd4dad9b7c/src/librustc_lint/bad_style.rs#L108) and not in clippy, so that would make it an issue in rust itself.

I would like to work on that because the linting code seems somewhat understandable, but since the [impl Trait feature](https://github.com/rust-lang/rust/issues/34511) is still going through some work, I'm not really sure if it makes sense to fix something that might be changed later?

**Comment by kimsnj** [comment]
Hi,

I tried having a look at this issue, but I'm unable to reproduce it.
I used the following example using both Rust Playground Clippy and using clippy-driver in local, but I get no warning from Clippy: 

```rust
#![feature(universal_impl_trait)]
#![feature(conservative_impl_trait)]

fn spy<'a>(x: impl Iterator<Item = &'a str>) -> impl Iterator<Item = &'a str> {
    x.map(|s| {
        println!("- {}", s);
        s
    })
}

fn main() {
    let s = "one\ntwo\nthree";
    let _lines: Vec<_> = spy(s.lines()).collect();
}
```
Actually (rip)grepping `non_camel_case_types` in the repo didn't return any result.

Did I miss something ?

**Comment by oli-obk** [comment]
Indeed this is a rustc error. It would still make sense to fix it in rustc, because this is something that will be important irrelevant of the final design questions of impl Trait
